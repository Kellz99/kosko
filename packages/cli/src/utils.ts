export function excludeFalsyInArray<T>(input: (T | undefined | null)[]): T[] {
  return input.filter(Boolean) as T[];
}
