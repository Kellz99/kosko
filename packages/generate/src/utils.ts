export function validateConcurrency(value = 10): number {
  if (value < 1) {
    throw new Error("Concurrency must be greater than 0");
  }

  return value;
}
