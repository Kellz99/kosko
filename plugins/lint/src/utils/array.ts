export function difference<T>(a: T[], b: T[]): T[] {
  const set = new Set(b);
  return a.filter((x) => !set.has(x));
}
