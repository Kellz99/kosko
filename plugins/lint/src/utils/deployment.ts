import type { IPodTemplateSpec } from "kubernetes-models/v1/PodTemplateSpec";
import type { ResourcePath } from "./manifest";
import type { PartialDeep } from "type-fest";
import { apiVersionToGroup, isRecord } from "@kosko/common-utils";
import type { ILabelSelector } from "@kubernetes-models/apimachinery/apis/meta/v1/LabelSelector";
import { getObjectValue } from "./object";

const deploymentLikeResources: ResourcePath[] = [
  {
    apiGroup: "",
    kind: "ReplicationController",
    keys: ["spec"]
  },
  {
    apiGroup: "apps",
    kind: "Deployment",
    keys: ["spec"]
  },
  {
    apiGroup: "apps",
    kind: "StatefulSet",
    keys: ["spec"]
  },
  {
    apiGroup: "apps",
    kind: "DaemonSet",
    keys: ["spec"]
  },
  {
    apiGroup: "apps",
    kind: "ReplicaSet",
    keys: ["spec"]
  },
  {
    apiGroup: "batch",
    kind: "Job",
    keys: ["spec"]
  },
  {
    apiGroup: "batch",
    kind: "CronJob",
    keys: ["spec", "jobTemplate", "spec"]
  },
  {
    apiGroup: "argoproj.io",
    kind: "Rollout",
    keys: ["spec"]
  },
  {
    apiGroup: "serving.knative.dev",
    kind: "Service",
    keys: ["spec"]
  },
  {
    apiGroup: "serving.knative.dev",
    kind: "Configuration",
    keys: ["spec"]
  }
];

export interface DeploymentLikeSpec {
  template?: PartialDeep<IPodTemplateSpec>;
  selector?: ILabelSelector;
}

export function getDeploymentLikeSpec(
  value: unknown
): DeploymentLikeSpec | undefined {
  if (!isRecord(value)) return;

  for (const resource of deploymentLikeResources) {
    if (
      typeof value.apiVersion === "string" &&
      apiVersionToGroup(value.apiVersion) === resource.apiGroup &&
      value.kind === resource.kind
    ) {
      const spec = getObjectValue(value, resource.keys);

      if (isRecord(spec)) {
        return spec;
      }
    }
  }
}
