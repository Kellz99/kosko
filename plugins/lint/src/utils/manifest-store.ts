import type { Manifest } from "../rules/types";
import { apiVersionToGroup } from "@kosko/common-utils";

export interface Predicate {
  apiVersion?: string;
  apiGroup?: string;
  kind?: string;
  namespace?: string;
  name?: string;
}

function getIndexKeys(manifest: Manifest): string[] {
  const metadata = manifest.metadata;
  if (!metadata) return [];

  const keys: string[] = [];

  if (typeof metadata.apiVersion === "string") {
    keys.push(`apiVersion=${metadata.apiVersion}`);
    keys.push(`apiGroup=${apiVersionToGroup(metadata.apiVersion)}`);
  }

  if (typeof metadata.kind === "string") {
    keys.push(`kind=${metadata.kind}`);
  }

  if (typeof metadata.namespace === "string") {
    keys.push(`namespace=${metadata.namespace}`);
  } else {
    keys.push("namespace=");
  }

  if (typeof metadata.name === "string") {
    keys.push(`name=${metadata.name}`);
  }

  return keys;
}

export class ManifestStore {
  private readonly index: Record<string, Set<number>> = {};

  public constructor(private readonly store: readonly Manifest[]) {
    for (let i = 0; i < store.length; i++) {
      for (const key of getIndexKeys(store[i])) {
        this.index[key] ??= new Set();
        this.index[key].add(i);
      }
    }
  }

  public find<T>(predicate: Predicate): Manifest<T> | undefined {
    let result = Array.from({ length: this.store.length }, (_, i) => i);

    for (const [key, value] of Object.entries(predicate)) {
      const indexKey = `${key}=${value ?? ""}`;
      const index = this.index[indexKey];

      if (!index) return;

      result = result.filter((i) => index.has(i));

      if (!result.length) return;
    }

    return this.store[result[0]] as Manifest<T>;
  }

  public forEach(callback: (value: Manifest) => void): void {
    this.store.forEach((value) => callback(value));
  }
}
