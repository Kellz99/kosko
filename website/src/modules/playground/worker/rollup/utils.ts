export function getModuleURLForCDN(module: string) {
  return `https://cdn.skypack.dev/${module}`;
}
