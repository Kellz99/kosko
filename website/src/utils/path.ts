export function isAbsolute(path: string): boolean {
  return path.startsWith("/");
}

export function isRelative(path: string): boolean {
  return path.startsWith("./") || path.startsWith("../");
}

export function dirname(path: string): string {
  const index = path.lastIndexOf("/");
  if (index === -1) return path;
  return path.substring(0, index - 1) || "/";
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf("/");
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split("/").filter(Boolean);
  const toParts = to.split("/").filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join("/");
}
