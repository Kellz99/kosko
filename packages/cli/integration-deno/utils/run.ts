import execa from "execa";
import { join } from "node:path";
import { projectRoot } from "@kosko/test-utils";

export async function runDenoCLI(
  args: string[],
  options: execa.Options = {}
): Promise<execa.ExecaReturnValue> {
  return execa(
    "deno",
    [
      "run",
      "-A",
      "--import-map",
      join(projectRoot, "packages/cli/templates/deno/import_map.json"),
      "npm:kosko",
      ...args
    ],
    {
      ...options,
      env: {
        LC_ALL: "en_US",
        ...options.env
      }
    }
  );
}
