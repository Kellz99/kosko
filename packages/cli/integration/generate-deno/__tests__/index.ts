import { dirname } from "node:path";
import { runDenoCLI } from "../../utils/run";

test("should generate components", async () => {
  const result = await runDenoCLI(["generate"], {
    cwd: dirname(__dirname)
  });

  expect(result.stdout).toMatchSnapshot();
});
