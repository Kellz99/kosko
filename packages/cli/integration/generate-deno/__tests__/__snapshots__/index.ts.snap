// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate components 1`] = `
"---
apiVersion: v1
kind: Pod
metadata:
  name: file-esm
---
apiVersion: v1
kind: Pod
metadata:
  name: folder-esm"
`;
